Support returned point types as params to later queries

craigtaverner opened the pull request:
The public types created earlier this year made it impossible to return point objects as parameters to other queries. This code fixes that by adding the reverse conversion of public types back to private when executing cypher. The code for converting in both directions is isolated to the same conversion object for future use when working on separating Cypher and kernel more.

changelog: Fixed inability to use returned point types as parameters to later queries

craigtaverner committed:
Support returned point types as params to later queries

craigtaverner commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;3snRNfaf&#x2F;11-cannot-pass-back-in-point-type-as-param"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Cannot pass back in point type as param</a>

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -19,17 +19,38 @@
+
+import scala.collection.JavaConverters._
+
+trait javaCompatGeom extends graphdb.spatial.Geometry {
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -19,17 +19,38 @@
 }
 
-trait Point extends Geometry {
+trait Point extends Geometry with graphdb.spatial.Point {
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -39,7 +60,11 @@ case class GeographicPoint(longitude: Double, latitude: Double, crs: CRS) extend
 }
 
-case class CRS(name: String, code: Int, url: String)
+case class CRS(name: String, code: Int, url: String) extends graphdb.spatial.CRS {
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -62,4 +87,40 @@ object CRS {
     case _ => throw new InvalidArgumentException(s"HREF '$url' does not match any supported coordinate reference system for points, supported CRS are: '${WGS84.url}', '${Cartesian.url}'")
   }
+
+  def fromExternal(crs: graphdb.spatial.CRS) = CRS(crs.getType, crs.getCode, crs.getHref)
```
this translation should live outside `v3_0`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -62,4 +87,40 @@ object CRS {
 }
+
+object Points {
+  def fromPublic(external: graphdb.spatial.Point): Point = {
```
this translation should live outside `v3_0`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -19,17 +19,38 @@
-trait Geometry {
-  def coordinates: Seq[Double]
+import org.neo4j.cypher.internal.frontend.v3_0.{CypherTypeException, InvalidArgumentException}
+import org.neo4j.graphdb
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/Geometry.scala:
```diff
@@ -19,17 +19,38 @@
-  def coordinates: Seq[Double]
+import org.neo4j.cypher.internal.frontend.v3_0.{CypherTypeException, InvalidArgumentException}
+import org.neo4j.graphdb
+import org.neo4j.graphdb.spatial.Coordinate
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/DistanceFunction.scala:
```diff
@@ -23,9 +23,13 @@ import java.lang.Math._
+import org.neo4j.cypher.internal.compiler.v3_0.{Points, Point, CRS, ExecutionContext}
 import org.neo4j.cypher.internal.frontend.v3_0.CypherTypeException
 import org.neo4j.cypher.internal.frontend.v3_0.symbols._
+import org.neo4j.graphdb.spatial.Coordinate
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/DistanceFunction.scala:
```diff
@@ -23,9 +23,13 @@ import java.lang.Math._
 import org.neo4j.cypher.internal.frontend.v3_0.CypherTypeException
 import org.neo4j.cypher.internal.frontend.v3_0.symbols._
+import org.neo4j.graphdb.spatial.Coordinate
+import org.neo4j.graphdb
```
`compiler.v3_0` should not depend on stuff in `org.neo4j.graphdb`

craigtaverner committed:
Support points created as geometry type

craigtaverner committed:
Support use of public types returned from procedures

In particular we support procedures that return the public Point type and this
can then be used in the distance function which accepts the internal Point type.
The conversion is done with a similar mapper principle to that introduced for
internal to public type conversions in 079e1bd6e6aec2440527b94a3bbc4482256eb58d.

craigtaverner committed:
Nicer scala collection mapping for type conversion

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/CypherCompiler.scala:
```diff
@@ -77,15 +77,16 @@ object CypherCompilerFactory {
                         runtimeName: Option[RuntimeName],
                         updateStrategy: Option[UpdateStrategy],
-                        publicTypeConverter: Any => Any): CypherCompiler = {
+                        publicTypeConverter: Any => Any,
```
maybe send in a 

``` scala
trait Converter {
   def convertPublic(in: Any): Any
   def convertPrivate(in: Any): Any
}
```

instead of passing in two different functions everywhere

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/DistanceFunction.scala:
```diff
@@ -35,20 +35,38 @@ case class DistanceFunction(p1: Expression, p2: Expression) extends Expression {
+  }
+
+  def calculateDistance(geometry1: Point, geometry2: Point) = {
+    Seq(
```
This list doesn't have to be created on each call

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/DistanceFunction.scala:
```diff
@@ -87,12 +105,12 @@ object HaversinCalculator extends DistanceCalculator {
-    val c1: Seq[Double] = p1.coordinates.map(toRadians)
-    val c2: Seq[Double] = p2.coordinates.map(toRadians)
+  override def calculateDistance(p1: Point, p2: Point): Double = {
+    val c1: Seq[Double] = p1.coordinate.values.map(toRadians)
```
`c1` and `c2` are accessed by index and should therefore either be `Array[Double]` or `Vector[Double]`

pontusmelke commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor3_0.scala:
```diff
@@ -56,14 +56,27 @@ object helpersv3_0 {
+  }
+
+  def asPrivateMap(incoming: Map[String, Any]): Map[String, Any] =
+    incoming.mapValues(asPrivateType)
```
I think you want to use `immutableMapValues(incoming, asDeepScalaValue)` here, `mapValues`  returns a lazy view of the map

craigtaverner committed:
Refactored due to PR comments

* Cleaner API for passing converters around
* Micro optimizations in distance function
* Removed asymmetry in Point/Geometry type conversion
* Support 'point as geometry' in distance function
